actors/builtin/storage_market/multiset.go

```diff
@@ -0,0 +1,127 @@
+package storage_market
+
+import (
+	"encoding/binary"
+
+	cid "github.com/ipfs/go-cid"
+	errors "github.com/pkg/errors"
+	cbg "github.com/whyrusleeping/cbor-gen"
+
+	"github.com/filecoin-project/specs-actors/actors/util/adt"
+)
+
+type SetMultimap struct {
+	mp *adt.Map
+	store adt.Store
+}
+
+// Interprets a store as a HAMT-based map of HAMT-based sets with root `r`.
+func AsSetMultimap(s adt.Store, r cid.Cid) *SetMultimap {
+	return &SetMultimap{mp: adt.AsMap(s, r), store: s}
+}
+
+// Creates a new map backed by an empty HAMT and flushes it to the store.
+func MakeEmptySetMultimap(s adt.Store) (*SetMultimap, error) {
+	m, err := adt.MakeEmptyMap(s)
+	return &SetMultimap{m, s}, err
+}
+
+// Returns the root cid of the underlying HAMT.
+func (mm *SetMultimap) Root() cid.Cid {
+	return mm.mp.Root()
+}
+
+func (mm *SetMultimap) Put(key adt.Keyer, idx uint64) error {
+	// Load the hamt under key, or initialize a new empty one if not found.
+	set, found, err := mm.get(key)
+	if err != nil {
+		return err
+	}
+	if !found {
+		set, err = adt.MakeEmptySet(mm.store)
+		if err != nil {
+			return errors.Wrapf(err, "failed to initialize multiset set value under root %v", mm.mp.Root())
+		}
+	}
+
+	// Add to the set.
+	if err = set.Put(adt.IntKey(idx)); err != nil {
+		return errors.Wrapf(err, "failed to add multiset key %v", key)
+	}
+
+	// Store the new set root under key.
+	newArrayRoot := cbg.CborCid(set.Root())
+	err = mm.mp.Put(key, &newArrayRoot)
+	if err != nil {
+		return errors.Wrapf(err, "failed to store multiset array")
+	}
+	return nil
+}
+
+// Removes a value for a key.
+func (mm *SetMultimap) Remove(key adt.Keyer, idx uint64) error {
+	// Load the array under key, or initialize a new empty one if not found.
+	set, found, err := mm.get(key)
+	if err != nil {
+		return err
+	}
+	if !found {
+		return adt.ErrNotFound{
+			Root: set.Root(),
+			Key:  []interface{}{key, idx},
+		}
+	}
```

Comment:
Is this worth erroring? We could just change the semantics to ensure it's not present afterwards? Else return a boolean?